### PR TITLE
Implementation of hashlink creation functionality.

### PR DESCRIPTION
**Pull request by msporny**
Initial implementation of the create functionality.

### CODE REVIEW

File: Hashlink.js [hunk 1]
```diff
@@ -23,8 +24,8 @@ export class Cipher {
    * is provided, the hashlink is created from the data.
    *
    * @param {Object} options - The options for the create operation.
-   * @param {Array} options.data - The data associated with the given URL. If
-   *   provided, this data is used to create the cryptographic hash.
+   * @param {Uint8Array} options.data - The data associated with the given URL.
```
Comment (dlongley) [inline Hashlink.js]: I think this is how you do optional params:

```suggestion
   * @param {Uint8Array} [options.data] - The data associated with the given URL.
```

File: Hashlink.js [hunk 2]
```diff
@@ -34,8 +35,65 @@ export class Cipher {
...
    */
-  async create({data, urls, transforms, meta}) {
-    throw new Error('Not implemented.');
+  async create({data, urls, transforms, meta = {}}) {
+    // ensure data or urls are provided
+    if(data === undefined && urls == undefined) {
+      throw new Error('Either `data` or `urls` must be provided.')
+    }
+
+    // transforms are provided
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
    // ensure transforms are provided
```

File: Hashlink.js [hunk 3]
```diff
@@ -34,8 +35,65 @@ export class Cipher {
...
+    if(transforms === undefined) {
+      throw new Error('The hashlink creation `transforms` must be provided.')
+    }
+
+    if(urls) {
+      // ensure urls are an array
+      if(Array.isArray(urls) === false) {
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
      if(!Array.isArray(urls)) {
```

File: Hashlink.js [hunk 4]
```diff
@@ -34,8 +35,65 @@ export class Cipher {
...
+        urls = [urls];
+      }
+
+      // ensure all URLs are strings
+      urls.forEach(url => {
+        if(typeof url !== 'string') {
+          throw new Error('Each `url` must be a string. Value: ' + url, url);
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
          throw new Error(`Url "${url}" must be a string.`);
```

File: Hashlink.js [hunk 5]
```diff
@@ -34,8 +35,65 @@ export class Cipher {
...
+        }
+      });
+    }
+
+    // merge meta options with urls
+    meta = Object.assign(meta, {'url': urls});
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
    meta = {...meta, url: urls};
```

Also, move this inside the `if(urls)` conditional above.

File: Hashlink.js [hunk 6]
```diff
@@ -34,8 +35,65 @@ export class Cipher {
...
+
+    // generate the encoded cryptographic hash
+    const outputData = await transforms.reduce(
+      async (transformedData, transform) => {
+      transformedData = await transformedData;
+      transformedData = await this.registeredTransforms[transform].encode(
+        transformedData);
+      return transformedData;
```
Comment (dlongley) [inline Hashlink.js]: Could shorten to:

```js
const outputData = await transforms.reduce(async (transformedData, transform) =>
  this.registeredTransforms[transform].encode(await transformedData), data);
```

File: Hashlink.js [hunk 7]
```diff
@@ -34,8 +35,65 @@ export class Cipher {
...
+    if(metadata.size > 0) {
+      const baseEncodingTransform = transforms[transforms.length - 1];
+      const cborData = new Uint8Array(cbor.encode(metadata));
+      const mbCborData = new TextDecoder().decode(
+        this.registeredTransforms[baseEncodingTransform].encode(cborData));
+
+      mhMeta = ':' + mbCborData;
+    }
+
+    return 'hl:' + new TextDecoder().decode(outputData) + mhMeta;
```
Comment (dlongley) [inline Hashlink.js]: Could reformulate this as:

```js
    const textDecoder = new TextDecoder();
    let hashlink = 'hl:' + textDecoder.decode(outputData);

    // append meta data if present
    if(metadata.size > 0) {
      const baseEncodingTransform = transforms[transforms.length - 1];
      const cborData = new Uint8Array(cbor.encode(metadata));
      const mbCborData = textDecoder.decode(
        this.registeredTransforms[baseEncodingTransform].encode(cborData));
      hashlink += ':' + mbCborData;
    }

    return hashlink;
```

File: Hashlink.js [hunk 8]
```diff
@@ -72,12 +130,12 @@ export class Cipher {
    * mechanisms such as new cryptographic hashing, base-encoding, and
    * resolution mechanisms.
    *
-   * @param {string} algorithm - An algorithm identifier.
-   * @param {Function} transform - A Function(input) that transforms input
-   *   data to output data.
+   * @param {Transform} transform - A Transform class that has a .encode()
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
   * @param {Transform} transform - A Transform instance that has a .encode()
```

File: Hashlink.js [hunk 9]
```diff
@@ -72,12 +130,12 @@ export class Cipher {
...
+   *   and a .decode() method. It must also have an ```identifier``` and
+   *   ```algorithm``` property.
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
   *   `algorithm` property.
```

File: Hashlink.js [hunk 10]
```diff
@@ -72,12 +130,12 @@ export class Cipher {
    * mechanisms such as new cryptographic hashing, base-encoding, and
    * resolution mechanisms.
    *
-   * @param {string} algorithm - An algorithm identifier.
-   * @param {Function} transform - A Function(input) that transforms input
-   *   data to output data.
+   * @param {Transform} transform - A Transform class that has a .encode()
+   *   and a .decode() method. It must also have an ```identifier``` and
```
Comment (dlongley) [inline Hashlink.js]:
```suggestion
   *   and a .decode() method. It must also have an `identifier` and
```

**Review by dlongley — CHANGES_REQUESTED** [review]

File: Hashlink.js (hunk 6 above)
Comment (msporny) [reply Hashlink.js]: Done in ebd661c1fcaceff5eccd5269caa818ff134c9af5

File: Hashlink.js (hunk 7 above)
Comment (msporny) [reply Hashlink.js]: Simplified in ebd661c1fcaceff5eccd5269caa818ff134c9af5

File: Hashlink.js (hunk 9 above)
Comment (msporny) [reply Hashlink.js]: Done in ebd661c1fcaceff5eccd5269caa818ff134c9af5.

**Pull request merged**